### PR TITLE
Retries for "Jupyter kernel is not ready" error

### PR DESCRIPTION
**Pull request by akarukappadath**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala [hunk 1]
```diff
@@ -481,12 +481,15 @@ trait LeonardoTestUtils extends WebBrowserSpec with Matchers with Eventually wit
 
   def withNewNotebook[T](cluster: Cluster, kernel: Kernel = Python2, timeout: FiniteDuration = 2.minutes)(testCode: NotebookPage => T)(implicit webDriver: WebDriver, token: AuthToken): T = {
     withNotebooksListPage(cluster) { notebooksListPage =>
-      notebooksListPage.withNewNotebook(kernel, timeout) { notebookPage =>
-        testCode(notebookPage)
```
Comment (kyuksel) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala]: The retry timeout below is `5 minutes`. However, the enclosing `withNewNotebook`'s `timeout` parameter defaults to `2 minutes` (which tests don't typically override) which should make the retry timeout effectively `2 minutes` as well. If we need it to be `5 minutes`, we may want to update the enclosing one. If not, let's lower it to `2 minutes` so it's not misleading.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala [hunk 2]
```diff
@@ -481,12 +481,15 @@ trait LeonardoTestUtils extends WebBrowserSpec with Matchers with Eventually wit
...
+      retryUntilSuccessOrTimeout(failureLogMessage = s"Cannot make new notebook")(10 seconds, 5 minutes) {() =>
+        notebooksListPage.withNewNotebook(kernel, timeout) { notebookPage =>
+          testCode(notebookPage)
+        }
       }
     }
   }
 
+
```
Comment (kyuksel) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala]: Nit: remove the redundant line above

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala (hunk 1 above)
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala]: The `retryUntilSuccessOrTimeout` method is a little misleading. If you look at the code:
```
def retryUntilSuccessOrTimeout[T](pred: Predicate[Throwable] = always, failureLogMessage: String = defaultErrorMessage)(interval: FiniteDuration, timeout: FiniteDuration)(op: () => Future[T])(implicit executionContext: ExecutionContext): RetryableFuture[T] = {
  val trialCount = Math.ceil(timeout / interval).toInt
  retryInternal(Seq.fill(trialCount)(interval), pred, failureLogMessage)(op)
}
```
it doesn't actually enforce the `timeout` passed to it, it just uses the interval/timeout to determine the number of tries. So this code as written will try up to 4 times, sleeping 30 seconds between each try. Each try will take at most 2 minutes. So at most, this test will take 4*2m + (0.5m*3) = 9.5 minutes. (Hopefully I got that right..)

I think the current values are probably appropriate, just giving more details about how this works.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala [hunk 3]
```diff
@@ -481,8 +481,10 @@ trait LeonardoTestUtils extends WebBrowserSpec with Matchers with Eventually wit
 
   def withNewNotebook[T](cluster: Cluster, kernel: Kernel = Python2, timeout: FiniteDuration = 2.minutes)(testCode: NotebookPage => T)(implicit webDriver: WebDriver, token: AuthToken): T = {
     withNotebooksListPage(cluster) { notebooksListPage =>
-      notebooksListPage.withNewNotebook(kernel, timeout) { notebookPage =>
-        testCode(notebookPage)
+      retryUntilSuccessOrTimeout(failureLogMessage = s"Cannot make new notebook")(30 seconds, 2 minutes) {() =>
+        notebooksListPage.withNewNotebook(kernel, timeout) { notebookPage =>
```
[the remark this answers is not in the document]
Comment (kyuksel) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala]: Oh yea...thanks for catching that!

**Comment by akarukappadath** [comment]
jenkins retest

**Pull request merged**